src/Common/Core/Test/Services/ServiceManagerTests.cs

```diff
@@ -42,26 +36,74 @@ public class ServiceManagerTests {
         }
 
         [Test]
-        public void DoubleAdd() {
-            _serviceManager.AddService<I1>(new C1());
-            
-            Action a = () => _serviceManager.AddService<I1>(new C1());
-            a.ShouldThrow<InvalidOperationException>();
-        }
-
-        [Test]
         public void AddRemove() {
             var s = new C1();
             _serviceManager.AddService(s);
             _serviceManager.RemoveService(s);
             _serviceManager.GetService<I1>().Should().BeNull();
 
-            _serviceManager.AddService(s as I1);
+            _serviceManager.AddService((I1) s);
             _serviceManager.RemoveService<I1>();
             _serviceManager.GetService<I1>().Should().BeNull();
         }
 
         [Test]
+        public void AddRemoveDerived() {
+            var d = new Derived(_serviceManager);
+            _serviceManager.GetService<Derived>().Should().NotBeNull();
+            _serviceManager.GetService<Base>().Should().NotBeNull();
+            _serviceManager.GetService<I1>().Should().NotBeNull();
+            _serviceManager.GetService<I2>().Should().NotBeNull();
+
+            d.Dispose();
+            _serviceManager.GetService<Derived>().Should().BeNull();
+            _serviceManager.GetService<Base>().Should().BeNull();
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy01() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<I2>();
+
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy02() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<I1>();
+
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy03() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<Base>();
```

Comment:
I don't like this idea, do we really need it? It makes service removal operation more accessible than adding: to add a service, code needs access to the type itself or one of its constructors. To remove a type, code needs access to any of the interfaces this type implements. And in case when several types implement one and the same interface, there is no way to find out what service will be removed. 

This can be an unpleasant bug prone. For example, imagine the following code:
```cs
public class Service : IDisposable { /*...*/ }

public class ServiceProvider : IDisposable {
    private IServiceManager _services;
    private Service _service;
    public Client(IServiceManager services) {
        _services = services;
        _service = new Service();
        services.AddService(_service);
    }

    public void Dispose() {
        _services.Remove(_service);
        _service.Dispose();
    }
}
```

It will work as expected only until `_service` field is of `Service`. If someone changes it to `IDisposable`, behavior will be broken (cause there is more than one service that implements `IDisposable`) and it will be hard to find out why.